scopesim/source/source_fields.py

```diff
@@ -0,0 +1,314 @@
+reference for each individual point source, ``ImageSourceField`` for a 2D image
+with an average spectrum, and finally ``CubeSourceField`` with a full 3D data
+cube. The ``ImageSourceField`` and ``CubeSourceField`` also contain a ``WCS``
+coordinate information and the wavelength axis of the ``CubeSourceField`` is
+available via the ``CubeSourceField.wave`` attribute.
+
+In previous versions of ScopeSim (pre-0.9), the ``Source.fields`` collection
+simply held the individual ``Table`` and ``ImageHDU`` (2D or 3D) objects, which
+are now stored in the ``.field`` attribute of each source field. This new
+distinction of the different cases allows much clearer separation of the logic
+required to handle various operations on those objects, such as plotting and
+shifting the source, which previously had to incorporate a number of case
+differentiations that made the ``Source`` class rather oberloaded with logic.
+This now also allows for well-structured validation logic of the individual
+source field data upon creation of each ``SourceField`` subclass instance.
+
+Creation of the source field classes is usually handled by the ``Source`` class
+itself via its various constructions methods, so the user rarely interacts with
+these classes directly, except for debugging. They serve more as an internal
+abstraction layer to handle the different cases of target object descriptions,
+as described above.
+
+The following class diagramm illustrates the relationship between the
+``SourceField`` subclasses:
+
+```mmd
+classDiagram
+class SourceField{+field}
+class SpectrumSourceField{+spectra}
+class HDUSourceField{+wcs}
+
+SourceField <|-- SpectrumSourceField
+SourceField <|-- HDUSourceField
+SpectrumSourceField <|-- TableSourceField
+SpectrumSourceField <|-- ImageSourceField
+HDUSourceField <|-- ImageSourceField
+HDUSourceField <|-- CubeSourceField
+```
+
+"""
+
+from copy import deepcopy
+from pathlib import Path
+from typing import TextIO, Any
+from dataclasses import dataclass, KW_ONLY, field as dataclass_field
+# rename it dataclass_field to avoid confusion with source field
+
+import numpy as np
+
+from astropy.table import Table, Column
+from astropy.io.registry import IORegistryError
+from astropy.io import fits
+from astropy import units as u
+from astropy.wcs import WCS, SingularMatrixError, FITSFixedWarning
+
+from synphot import SourceSpectrum
+
+
+from ..optics import image_plane_utils as imp_utils
+from ..utils import (quantify, quantity_from_table, close_loop, get_logger,
+                     convert_table_comments_to_dict)
+
+
+logger = get_logger(__name__)
+
+
+# TODO: consider making this a metaclass
+@dataclass
+class SourceField:
+    """Base class for source fields, not meant to be instantiated."""
+
+    field: Any
+    _: KW_ONLY
+    meta: dict = dataclass_field(default_factory=dict, repr=False)
+
+    def _write_stream(self, stream: TextIO) -> None:
+        raise NotImplementedError("Subclasses should implement this.")
```

Comment:
See comment on `Source.__repr__()`. I believe `_write_stream()` is not necessary.